### PR TITLE
Tg 125: React Hook Forms to the Employee tab

### PR DESCRIPTION
**Pull request by FefeTroanes**
- Loader component implemented to the Employee tab.
- React Hook Forms added to the Employee's add and edit forms, so now they have validations and error messages are showed when you blur the inputs.
- Relative imports used in the Employees tab changed to absolute imports.
- Code splitted for the employees so the initial load should be faster.

### CODE REVIEW

File: src/Components/Employees/index.jsx [hunk 1]
```diff
@@ -1,13 +1,42 @@
...
+import Dropdown from 'Components/Shared/Dropdown';
+import Form from 'Components/Shared/Form';
+import Button from 'Components/Shared/Button';
+import * as thunks from 'redux/employees/thunks';
 import { useDispatch, useSelector } from 'react-redux';
+import { useForm } from 'react-hook-form';
+import Joi from 'joi';
+import { joiResolver } from '@hookform/resolvers/joi';
+import Loader from 'Components/Shared/Loader';
+// import Input from 'Components/Shared/Input';
```
Comment (martinalbrnz) [inline src/Components/Employees/index.jsx:15]: ojo el comment

File: src/Components/Employees/index.jsx [hunk 2]
```diff
@@ -1,13 +1,42 @@
...
+  gender: Joi.string().valid('Male', 'Female', 'Other'),
+  address: Joi.string().regex(/^[a-zA-Z0-9\s,'-]*$/),
+  dob: Joi.date().required(),
+  password: Joi.string()
+    .min(8)
+    .required()
+    .regex(/(?!^[0-9]*$)(?!^[a-zA-Z]*$)^([a-zA-Z0-9]{8,25})$/),
+  phone: Joi.string().regex(/^[0-9\-+]{9,10}$/),
+  active: Joi.boolean()
+});
```
Comment (martinalbrnz) [inline src/Components/Employees/index.jsx:39]: Cuando haces el schema podes personalizar los mensajes que vas a recibir con cadauna de las validaciones de la siguiente forma
```
description: Joi.string()
      .min(3)
      .max(200)
      .messages({
        'string.min': 'Description must contain 3 or more characters',
        'string.max': 'Name must contain 200 or less characters',
        'string.empty': 'This field is required'
      })
      .required(),
```

Para esto, fijate como son los type de los errores y vas completando todos.
Para ver los errores, simplemente pones un `console.log(errors);` en algun lado del codigo y vas haciendo que fallen las validaciones de a una por vez.

File: src/Components/Employees/index.jsx [hunk 3]
```diff
@@ -88,52 +114,45 @@ function Employees() {
...
-    setActive(employeeToEdit[0].active === 'true' ? true : false);
     setShowAddEdit(true);
     setEmployeeIdToEdit(id);
+    reset({
+      firstName: employeeToEdit[0].firstName,
+      lastName: employeeToEdit[0].lastName,
+      email: employeeToEdit[0].email,
+      gender: employeeToEdit[0].gender,
+      address: employeeToEdit[0].address,
+      dob: new Date(employeeToEdit[0].dob).toISOString().split('T')[0] || '',
```
Comment (martinalbrnz) [inline src/Components/Employees/index.jsx:125]: Nice one

File: src/Components/Employees/index.jsx [hunk 4]
```diff
@@ -88,52 +114,45 @@ function Employees() {
...
+      password: '',
+      phone: '',
+      active: false
+    });
   };
 
   const closeForm = () => {
     setShowAddEdit(false);
     setMethod('');
+    resetFields();
```
Comment (martinalbrnz) [inline src/Components/Employees/index.jsx:149]: Nice one

File: src/Components/Employees/index.jsx [hunk 5]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
           </div>
+          {/* <Input
+            type={'text'}
+            htmlForProp={'First name'}
+            label={'First name'}
+            register={register}
+            name={'firstName'}
+            error={errors.firstName?.message}
+            width={'189.63px'}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:241]: borra estos coments si estan de mas

File: src/Components/Employees/index.jsx [hunk 6]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.lastName && <p className={styles.error}>{errors.lastName.message}</p>}
           </div>
+          {/* <Input
+            type={'text'}
+            htmlForProp={'Last name'}
+            label={'Last name'}
+            register={register}
+            name={'lastName'}
+            error={errors.lastName?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:254]: idem

File: src/Components/Employees/index.jsx [hunk 7]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.email && <p className={styles.error}>{errors.email.message}</p>}
           </div>
+          {/* <Input
+            type={'text'}
+            htmlForProp={'Email'}
+            label={'Email'}
+            register={register}
+            name={'email'}
+            error={errors.email?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:267]: idem

File: src/Components/Employees/index.jsx [hunk 8]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.address && <p className={styles.error}>{errors.address.message}</p>}
           </div>
+          {/* <Input
+            type={'text'}
+            htmlForProp={'Address'}
+            label={'Address'}
+            register={register}
+            name={'address'}
+            error={errors.address?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:295]: idem

File: src/Components/Employees/index.jsx [hunk 9]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.dob && <p className={styles.error}>{errors.dob.message}</p>}
           </div>
+          {/* <Input
+            type={'date'}
+            htmlForProp={'Date of birth'}
+            label={'Date of birth'}
+            register={register}
+            name={'dob'}
+            error={errors.dob?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:308]: idem

File: src/Components/Employees/index.jsx [hunk 10]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.password && <p className={styles.error}>{errors.password.message}</p>}
           </div>
+          {/* <Input
+            type={'password'}
+            htmlForProp={'Password'}
+            label={'Password'}
+            register={register}
+            name={'password'}
+            error={errors.password?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:321]: idem

File: src/Components/Employees/index.jsx [hunk 11]
```diff
@@ -203,98 +219,125 @@ function Employees() {
...
+            {errors.phone && <p className={styles.error}>{errors.phone.message}</p>}
           </div>
+          {/* <Input
+            type={'number'}
+            htmlForProp={'Phone'}
+            label={'Phone'}
+            register={register}
+            name={'phone'}
+            error={errors.phone?.message}
+          /> */}
```
Comment (nicolascostanza) [inline src/Components/Employees/index.jsx:334]: idem

File: src/Components/Employees/index.jsx (hunk 3 above)
Comment (FefeTroanes) [reply src/Components/Employees/index.jsx:125]: Eso lo hizo gina me parece 😝

**Review by nicolascostanza — COMMENTED** [review]
solo borrar esas cosas ademas de los comments de martin y ya estaria

File: src/Components/Employees/index.jsx (hunk 2 above)
Comment (FefeTroanes) [reply src/Components/Employees/index.jsx:39]: Dale, ahora los agrego

**Review by martinalbrnz — DISMISSED** [review]
LGTM

**Comment by Clarosluciano** [comment]
# Test result
Functionalities (CRUD): pass
Validations: fail
Localy tested on: https://alfonso-trackgenix-es0j4fcdx-basp-m2022.vercel.app/employees

# Bugs found:
1. The validation length is not triggered, only the validation "this field is required" appears:
image
image
**Expected behavior:** The validation length warns the user that the field has to be filled correctly.

2. The dob input can be filled with dates that have not yet passed:
image
**Expected behavior:** A validation should be triggered on that input.

3. Successful message is not centered:
image
image
image
**Expected behavior:** A symmetrical visual flow should be followed that does not clash with the rest of the modal.

4. There is no visible separation between the confirmation buttons:
image
**Expected behavior:** Buttons do not interfere with each other and are spaced at a pleasing distance.

**Review by martinalbrnz — APPROVED** [review]
Good job!

**Review by nicolascostanza — APPROVED** [review]

**Pull request merged**